### PR TITLE
Add skills system with Claude Code compatibility

### PR DESCRIPTION
**Pull request by nicobailon**
This addresses #169

Skills are now auto-discovered and listed in the system prompt. When a task matches a skill's description, the agent loads it with the read tool.

Scans these locations (later ones win on name collision): 
- `~/.claude/skills/*/SKILL.md` 
- `<cwd>/.claude/skills/*/SKILL.md` 
- `~/.pi/agent/skills/*.md` 
- `<cwd>/.pi/skills/*.md`

Claude Code's `allowed-tools` and `model` fields are ignored. The `{baseDir}` placeholder is handled by telling the agent each skill's base directory path in the prompt, so it can substitute when reading.

Disable with `--no-skills` or `skills.enabled: false` in settings.

@badlogic thoughts? Anything else you'd do differently?

### CODE REVIEW

File: packages/coding-agent/src/core/skills/index.ts [hunk 1]
```diff
@@ -0,0 +1,2 @@
+export { loadSkills } from "./loader.js";
```
Comment (badlogic) [inline packages/coding-agent/src/core/skills/index.ts]: I don't think we need a subfolder core/skills/, index.ts and types.ts are trivial, and the types can be at the start of skills.ts. Let's put skills.ts in core/ directly with the types at the front.

File: packages/coding-agent/src/core/skills/loader.ts [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+			} else if (key === "description") {
+				frontmatter.description = value;
+			}
+		}
+	}
+
+	return { frontmatter, body };
+}
+
+function loadPiSkillsFromDir(dir: string, source: SkillSource, subdir: string = ""): Skill[] {
```
Comment (badlogic) [inline packages/coding-agent/src/core/skills/loader.ts]: Why do we have separate loadPiSkillsFromDir and loadClaudeSkillsFromDir? Shouldn't the logic be exactly the same, with only the source differing?

File: packages/coding-agent/src/core/system-prompt.ts [hunk 3]
```diff
@@ -101,12 +102,37 @@ export function loadProjectContextFiles(): Array<{ path: string; content: string
 	return contextFiles;
 }
 
+function buildSkillsSection(skills: Skill[]): string {
+	if (skills.length === 0) {
+		return "";
+	}
+
+	const lines = [
```
Comment (badlogic) [inline packages/coding-agent/src/core/system-prompt.ts:110]: We should probably follow the "syntax" Anthropic are using for their skills tool in CC. Specifically, wrap the skills in these tags.

```
<available_skills>

</available_skills>
```

File: packages/coding-agent/docs/skills.md [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+```
+
+### Frontmatter Fields
+
+| Field | Required | Description |
+|-------|----------|-------------|
+| `description` | Yes | Short description for skill selection |
+| `name` | No | Override skill name (defaults to filename or directory name) |
+
+The parser only supports single-line `key: value` syntax. Multiline YAML blocks are not supported.
```
Comment (badlogic) [inline packages/coding-agent/docs/skills.md:42]: We should probably check out the skills Anthropic has up in this repo and see what the frontmatter looks like. I'm a little afraid of the frontmatter parser just breaking and exploding.

https://github.com/anthropics/skills

**Comment by badlogic** [comment]
@nicobailon great, just a few minor nits.

**Comment by nicobailon** [comment]
@badlogic Thanks for the call outs.

Flattened everything into core/skills.ts and merged the two load functions into one with a format param. Kept the branching internally since Pi and Claude skills do have different structures (flat .md files vs */SKILL.md directories), but it's all in one place now.

Added the `<available_skills>` tags.

For the parser concern - I cloned the Anthropic skills repo and tested against all 17 of their SKILL.md files. All pass. They're using simple single-line frontmatter, nothing fancy that would break us.

**Comment by nicobailon** [comment]
@badlogic went ahead and added Codex CLI skills support. They scan `~/.codex/skills/` recursively for `SKILL.md` files, same frontmatter format as Claude Code.

Put Codex at lowest priority so Claude/Pi skills win on name collision. Also added hidden entry and symlink skipping for all formats, which Codex does.

Let me know what you think!

**Comment by nicobailon** [comment]
@badlogic For reference, here's the Codex CLI commit that added skills support:
https://github.com/openai/codex/commit/a8d5ad37b8115068ff777b66ab9299cdab85cdd7

Their approach is basically identical to Claude Code - YAML frontmatter with name and description. Main difference is they scan recursively (`**/SKILL.md`) rather than one level deep (`*/SKILL.md`).

What I added:
- New "codex" format that recurses into all subdirectories looking for `SKILL.md` files
- Falls back to parent directory name if name field is missing (matching their behavior)
- Hidden entry (`. prefix`) and symlink skipping for all skill formats
- `~/.codex/skills/` loaded first so it has lowest priority

Full priority order:
1. `~/.codex/skills/**/SKILL.md` (recursive, lowest)
2. `~/.claude/skills/*/SKILL.md` (one level)
3. `<cwd>/.claude/skills/*/SKILL.md` (one level)
4. `~/.pi/agent/skills/**/*.md` (recursive)
5. `<cwd>/.pi/skills/**/*.md` (recursive, highest)

Tested by creating skills in `~/.codex/skills/test-codex/SKILL.md` (top-level), `~/.codex/skills/nested/deep/SKILL.md` (deeply nested), and `~/.codex/skills/.hidden/SKILL.md` (hidden). Pi found the first two and correctly skipped the hidden one.

**Comment by badlogic** [comment]
## Analysis: Skills System Implementation

Thanks for the comprehensive implementation and responsiveness to feedback. Here's my analysis:

### ✅ Strengths

**1. Clean Resolution of Review Feedback**
- Successfully flattened into single `core/skills.ts` file
- Merged separate load functions into unified `loadSkillsFromDir(dir, source, format)` 
- Added `<available_skills>` tags matching CC's syntax
- Tested against all 17 Anthropic SKILL.md files - all parse successfully

**2. Multi-Format Support**
- Pi: Recursive scanning with colon-separated names (`db:migrate`, `aws:s3:upload`)
- Claude Code: One-level directory structure (matches CC implementation)
- Codex CLI: Recursive scanning (matches Codex behavior)
- Smart prioritization: Codex < Claude < Pi (user control maintained)

**3. Security Considerations**
- Simple single-line YAML parser (`key: value` only) - no multiline complexity
- Skips hidden files and symlinks
- Graceful failure on invalid files (try/catch)
- No code execution (unlike bash expansion suggestions)

**4. Minimal Philosophy**
- Optional by default
- Toggleable: `--no-skills` CLI flag or `skills.enabled: false` in settings
- Only active when `read` tool available
- Aligns with "minimal, radical, optional" design

### 🤔 Areas for Consideration

**1. Placeholder Implementation**
Only `{baseDir}` currently supported. Should verify this covers all CC placeholders or document which ones are intentionally omitted.

**2. Silent Failures**
- Invalid frontmatter fails silently
- Name collisions have no warning (Pi skill overrides Claude skill with same name)
- Ignored `allowed-tools`/`model` fields have no runtime notification

Consider adding optional warnings or `--verbose` flag for debugging.

**3. Discovery Performance**
Scans 5 locations recursively on every startup. For users with large skill collections, could add:
- Lazy loading (scan on first skill use, not startup)
- Optional caching

**4. Testing & Validation**
Missing:
- Unit tests for frontmatter parser edge cases
- Tests for name collision behavior
- Discovery performance benchmarks
- Optional: `--list-skills` command for debugging

### 📋 Suggested Pre-Merge Checklist

**High Priority:**
- [ ] Add tests for frontmatter parsing (malformed YAML, missing description, edge cases)
- [ ] Document which CC placeholders are supported (confirm `{baseDir}` is complete list)

**Medium Priority:**
- [ ] Add `--list-skills` debug command
- [ ] Consider warning on name collisions
- [ ] Optional: `--verbose` flag for skill discovery debugging

**Low Priority:**
- [ ] Performance testing with 50+ skills
- [ ] Skill validation command (`pi check-skills`)

### 💡 Comparison to Current AGENTS.md Approach

**Your Current Manual Listing:**
- ✅ Simple, explicit, no magic
- ✅ Full control
- ❌ Manual maintenance
- ❌ No Claude Code compatibility
- ❌ All in context always

**This Skills System:**
- ✅ Auto-discovery - no maintenance
- ✅ Claude/Codex compatibility - lowers adoption barrier
- ✅ On-demand loading - only relevant instructions
- ✅ Project-specific skills support
- ❌ ~200 LOC complexity
- ❌ Discovery overhead
- ❌ Less explicit

### 🎯 Recommendation

**Approve after addressing high-priority items.** 

This is solid work that provides real value (CC compatibility) while maintaining Pi's philosophy. The code quality is good after refactoring, and nicopreme has been responsive to feedback.

The missing tests are the main blocker - frontmatter parsing has enough edge cases to warrant coverage. Once tests are added and placeholder support is documented, this is ready to merge.
